### ISSUE TITLE
Training

### ISSUE THREAD

**Issue by calm-focus** [digitalpathologybern/hover_next_train#2]
Thanks for the nice work. When traing by the code "torchrun --standalone --nnodes=1 --nproc-per-node=1 train.py --config "sample_configs/train_lizard.toml"", there is always a warning.

1  cuda devices
0
adding 3-class ground truth...
100%|█████████████████████████████████████| 4279/4279 [00:01<00:00, 2382.09it/s]
adding 3-class ground truth...
100%|███████████████████████████████████████| 702/702 [00:00<00:00, 2556.76it/s]
adding 3-class ground truth...
100%|█████████████████████████████████████| 2327/2327 [00:00<00:00, 2405.55it/s]
adding 3-class ground truth...
100%|█████████████████████████████████████| 1017/1017 [00:00<00:00, 2532.32it/s]
start step -1
inst_loss:  45.05625534057617 class_loss:  0.8221124410629272
/home/user/anaconda3/envs/hovernext/lib/python3.11/site-packages/torch/autograd/__init__.py:251: UserWarning: Grad strides do not match bucket view strides. This may indicate grad was not created according to the gradient layout contract, or that the param's strides changed since DDP was constructed.  This is not an error, but may impair performance.
grad.sizes() = [8, 16, 1, 1], strides() = [16, 1, 16, 16]
bucket_view.sizes() = [8, 16, 1, 1], strides() = [16, 1, 1, 1] (Triggered internally at ../torch/csrc/distributed/c10d/reducer.cpp:320.)
  Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass

I google this warining, people said .contiguous() should be after transpose or permute. After I did that, the warining still exists.

Can you help me with it?

**Comment by eliasbaumann** [comment]
Hey @calm-focus,

thanks for reaching out! Unfortunately pytorch has not yet made it easy to figure out what exactly causes this warning. It may be 1x1 conv, or some other specific operations. However generally you can ignore that warning since it does not actually prevent model convergence or training. For reference, for me on a single RTX 3090 training for the large model on lizard took ~3 days and for tiny pannuke ~1 day. 

Do you have any performance issues or was your concern mostly with that warning?

**Comment by calm-focus** [comment]
Thank you for your prompt reply. I retrained the model according to your configuration. The trained model segment and classfied cell well! Since I have 4 GPUs, can you suggest me any changes I need to make in the training parameters if I have 10,000k rather than 200k cells in Pannuke?

**Comment by eliasbaumann** [comment]
Hey @calm-focus, there are several adjustments you could make that the model could benefit from.

1. Increase the number of training_steps
2. Increase the combined batch size to the maximum possible on your GPUs
3. Reduce the learning rate
4. If you are certain about your labels, adjusting label_smoothing to a lower value might be beneficial
5. Adjust num_workers according to the number of available CPU cores per GPU (so if you have 64 cores for your 4GPU, maybe set it to 16 and see if that impacts training speed)

Also, make sure to optimize hyperparameters for the post-processing cutoffs, since your data is different than pannuke/lizard.

**Comment by calm-focus** [comment]
Thank you! @eliasbaumann

**Comment by calm-focus** [comment]
When using distributed training with multiple gpus, you set the same seed

random.seed(42)
dist.init_process_group("nccl")
torch.backends.cudnn.benchmark = True
torch.manual_seed(42)

for each gpu/rank. Does this lead to insufficient utilization of training data?  Maybe set for example 42 + int(os.environ["LOCAL_RANK"]) to let each gpu use diiferent batch for training for each iteration 
raw, gt = next(train_loaders[random.randint(0, len(train_loaders) - 1)])

**Comment by eliasbaumann** [comment]
Hey, 

agreed! I would move both random.seed as well as torch.manual_seed into the `supervised_training` function to give each worker a separate shuffle + augmentations.
I don't think the second part makes sense, because this is the part where you sample from one of the included datasets (and thats only relevant if you use more than one).
Maybe you could use distributed samplers for all dataloaders if you want to go that way.

Also since you are augmenting samples anyways, the different workers will see different samples at different times which essentially corresponds to including multiple augmentations of the same image during one epoch, which most likely will not make a difference over long training runs, so for an easier way to make sure that the workers are heterogeneous I would just include random.seed and torch.manual_seed in in the first few lines in supervised_training. but don't remove it from the main part.

**Comment by eliasbaumann** [comment]
Closing this issue due to inactivity, reopen it if there is anythine else related to this that you need!
